### PR TITLE
Reason for defaulting

### PR DESCRIPTION
**Pull request by paulinembabu**
See issue https://github.com/OpenSRP/opensrp-client-emmunize-malawi/issues/10

### CODE REVIEW

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java [hunk 1]
```diff
@@ -145,9 +156,23 @@ public boolean onOptionsItemSelected(MenuItem item) {
     @Override
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
+        AllSharedPreferences allSharedPreferences = getOpenSRPContext().allSharedPreferences();
+        try {
+            String jsonString = data.getStringExtra(JsonFormConstants.JSON_FORM_KEY.JSON);
+            Timber.d(jsonString);
+
+            JSONObject form = new JSONObject(jsonString);
+            String encounterType = form.getString(ChildJsonFormUtils.ENCOUNTER_TYPE);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java:165]: use a safer `optString` instead

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java [hunk 2]
```diff
@@ -145,9 +156,23 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+            if (encounterType.equalsIgnoreCase(GizConstants.EventType.REASON_FOR_DEFAULTING)) {
+                org.smartregister.child.util.Utils.startAsyncTask(new SaveReasonForDefaultingEventTask(jsonString, "", childDetails.entityId(), allSharedPreferences.fetchRegisteredANM(), CoreLibrary.getInstance().context().getEventClientRepository()), null);
 
+            }
+        } catch (JSONException e) {
+            e.printStackTrace();
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java:171]: Use `Timber`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java [hunk 3]
```diff
@@ -160,6 +166,9 @@ public void processClient(List<EventClient> eventClients) throws Exception {
                     processService(eventClient, serviceTable);
                 } else if (eventType.equals(ChildJsonFormUtils.BCG_SCAR_EVENT)) {
                     processBCGScarEvent(eventClient);
+                } else if (eventType.equals(GizConstants.EventType.REASON_FOR_DEFAULTING)) {
+                    Timber.e("creating Reason for DAFAULTING Here");
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java:170]: you can remove this now right or use `Timber.d`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java [hunk 4]
```diff
@@ -908,4 +917,19 @@ private void scheduleUpdatingClientAlerts(@NonNull String baseEntityId, @NonNull
             clientsForAlertUpdates.put(baseEntityId, dateTime);
         }
     }
+
+    private void clientProcessReasonForDefault(Event event) throws JsonProcessingException {
+        List<Obs> reasonForDefaultingObs = event.getObs();
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java:922]: Check for null

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                            cursor.getString(cursor.getColumnIndex(GizConstants.JsonAssetsHelper.OUTREACH_DEFAULTING_REASON)),
+                            cursor.getString(cursor.getColumnIndex(GizConstants.JsonAssetsHelper.ADDITIONAL_DEFAULTING_NOTES)),
+                            cursor.getString(cursor.getColumnIndex(GizConstants.JsonAssetsHelper.BASE_ENTITY_ID)),
+                            cursor.getString(cursor.getColumnIndex(GizConstants.JsonAssetsHelper.EVENT_DATE)),
+                            cursor.getString(cursor.getColumnIndex(GizConstants.JsonAssetsHelper.OTHER_DEFAULTING_REASON))));
+
+                } while (cursor.moveToNext());
+            }
+
+        } catch (Exception e) {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:85]: Catch a less general exception, line `SQLiteException` . Also why not use `try-with-resources`.

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            Timber.e(e.getMessage());
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:86]: You can leave it as `Timber.e(e)`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                cursor.close();
+            }
+        }
+        return reasonForDefaultingModels;
+    }
+
+    public void purgeReasonForDefaulting(String baseEntityID) {
+        try {
+            getWritableDatabase().execSQL("DELETE FROM reason_for_defaulting WHERE id ='" + baseEntityID + "'");
+        } catch (Exception e) {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:98]: Catch a less general exception, line `SQLiteException`.

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+import org.smartregister.giz.application.GizMalawiApplication;
+import org.smartregister.repository.EventClientRepository;
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Iterator;
+
+import timber.log.Timber;
+
+public class SaveReasonForDefaultingEventTask extends AsyncTask<Void, Void, Void> {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java:25]: Though we stopped using `AsyncTask`. Could we replace this with AppExecutors. What do you think?

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        this.providerId = providerId;
+        this.eventClientRepository = eventClientRepository;
+    }
+
+    @Override
+    protected Void doInBackground(Void... params) {
+        try {
+            processReasonForDefaultingEvent();
+        } catch (Exception e) {
+            Timber.e(Log.getStackTraceString(e));
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java:46]:
```suggestion
            Timber.e(e);
```
```suggestion
            Timber.e(Log.getStackTraceString(e));
```

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                .withEventDate(encounterDate)
+                .withEventType(encounterType)
+                .withLocationId(locationId)
+                .withProviderId(providerId).withEntityType(Constants.CHILD_TYPE)
+                .withChildLocationId(ChildLibrary.getInstance().context().allSharedPreferences().fetchCurrentLocality())
+                .withFormSubmissionId(ChildJsonFormUtils.generateRandomUUIDString()).withDateCreated(new Date());
+
+
+        //add metadata
+        addMetadata(fields, event, metadata);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java:84]: Why not use `JsonFormUtils.createEvent` from core.

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/GizReportUtils.java [hunk 11]
```diff
@@ -83,4 +92,70 @@ public static String getStringIdentifier(@NonNull String identifierCode) {
                 .replace(" ", "_")
                 .replace("/", "");
     }
+
+
+    public static ReasonForDefaultingModel getReasonForDefaultingRepository(List<Obs> defaultingObs) throws JsonProcessingException {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/GizReportUtils.java:97]: Is this method name right?, It sort of does something else.

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java [hunk 12]
```diff
@@ -283,6 +286,14 @@ public static synchronized GizMalawiApplication getInstance() {
         return vaccineGroups;
     }
 
+    public static String getObsValue(Obs obs) {
+        List<Object> values = obs.getValues();
+        if (values.size() > 0) {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java:291]: Check for null

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    public static void createTable(SQLiteDatabase database) {
+        database.execSQL(CREATE_TABLE_SQL);
+        database.execSQL(BASE_ID_INDEX);
+    }
+
+    public void addOrUpdate(ReasonForDefaultingModel reasonForDefaultingModel) {
+        addOrUpdateReasonForDefaulting(reasonForDefaultingModel, getWritableDatabase());
+    }
+
+    public void addOrUpdateReasonForDefaulting(ReasonForDefaultingModel reasonForDefaultingModel, SQLiteDatabase database) {
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:41]: Why add `database` as an argument?

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java (hunk 2 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java:171]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java (hunk 3 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/processor/GizMalawiProcessorForJava.java:170]: Removed

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java (hunk 5 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:85]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java (hunk 7 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:98]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java (hunk 8 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java:25]: Removed

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java (hunk 10 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/task/SaveReasonForDefaultingEventTask.java:84]: Replaced

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java (hunk 12 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/application/GizMalawiApplication.java:291]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/GizReportUtils.java (hunk 11 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/GizReportUtils.java:97]: Named Better

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java (hunk 13 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/repository/ReasonForDefaultingRepository.java:41]: Removed

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java (hunk 1 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/ChildDetailTabbedActivity.java]: Replaced

**Review by bennsimon — APPROVED** [review]

**Pull request merged**